man1/fhcfg.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "FHC\-FHCFG" "1" "December 2011" "" ""
.
.SH "NAME"
\fBfhc\-fhcfg\fR \- Manage the fhc Configuration File
.
.SH "SYNOPSIS"
.
.nf

fhc fhcfg set <key> <value> [\-\-global]
fhc fhcfg get <key>
fhc fhcfg delete <key>
fhc fhcfg list
fhc fhcfg edit
.
.fi
.
.SH "DESCRIPTION"
fhc gets its configuration values from 5 sources, in this priority:
.
.TP
cli
The command line flags\. Putting \fB\-\-foo bar\fR on the command line sets the \fBfoo\fR configuration parameter to \fB"bar"\fR\. A \fB\-\-\fR argument tells the cli parser to stop reading flags\. A \fB\-\-flag\fR parameter that is at the \fIend\fR of the command will be given the value of \fBtrue\fR\.
.
.TP
env
Any environment variables that start with \fBfhc_fhcfg_\fR will be interpreted as a configuration parameter\. For example, putting \fBfhc_fhcfg_foo=bar\fR in your environment will set the \fBfoo\fR configuration parameter to \fBbar\fR\. Any environment configurations that are not given a value will be given the value of \fBtrue\fR\. Config values are case\-insensitive, so \fBFHCFG_FOO=bar\fR will work the same\.
.
.TP
$HOME/\.fhcrc (or the \fBuserconfig\fR param, if set above)
This file is an ini\-file formatted list of \fBkey = value\fR parameters\.
.
.TP
$PREFIX/etc/fhcrc (or the \fBglobalconfig\fR param, if set above)
This file is an ini\-file formatted list of \fBkey = value\fR parameters
.
.TP
default configs
This is a set of configuration parameters that are internal to fhc, and are defaults if nothing else is specified\.
.
.SH "Sub\-commands"
fhcfg supports the following sub\-commands:
.
.SS "set"
.
.nf

fhc fhcfg set key value
.
.fi
.
.P
Sets the config key to the value\.
.
.P
If value is omitted, then it sets it to "true"\.
.
.SS "get"
.
.nf

fhc fhcfg get key
.
.fi
.
.P
Echo the config value to stdout\.
.
.SS "list"
.
.nf

fhc fhcfg list
.
.fi
.
.P
Show all the config settings\.
.
.SS "delete"
.
.nf

fhc fhcfg delete key
.
.fi
.
.P
Deletes the key from all configuration files\.
.
.SS "edit"
.
.nf

fhc fhcfg edit
.
.fi
.
.P
Opens the config file in an editor\. Use the \fB\-\-global\fR flag to edit the global config\.


lib/fhcfg.js

module.exports = config;

config.usage =   "fhc fhcfg set <key> <value>"
             + "\nfhc fhcfg get <key>"
             + "\nfhc fhcfg delete <key>"
             + "\nfhc fhcfg list"
             + "\nfhc fhcfg edit";

var ini = require("./utils/ini");
var log = require("./utils/log");
var fhc = require("./fhc");
var exec = require("./utils/exec");
var fs = require("fs");
var dc;
var output = require("./utils/output");
var types = require("./utils/config-defs").types;

// fhc config set key value
// fhc config get key
// fhc config list
function config (args, cb) {
  var action = args.shift();
  switch (action) {
    case "set": return set(args[0], args[1], cb);
    case "get": return get(args[0], cb);
    case "delete": case "rm": case "del": return del(args[0], cb);
    case "list": case "ls": return list(cb);
    case "edit": return edit(cb);
    default: return unknown(action, cb);
  }
}

function edit (cb) {
  var e = ini.get("editor")
    , f = ini.get(ini.get("global") ? "globalconfig" : "userconfig");
  if (!e) return cb(new Error("No EDITOR config or environ set."));
  ini.save(function (er) {
    if (er) return cb(er);
    fs.readFile(f, "utf8", function (er, data) {
      if (er) data = "";
      dc = dc || require("./utils/config-defs").defaults;
      data = [ ";;;;"
             , "; fhc "+(ini.get("global") ? "globalconfig" : "userconfig")+" file"
             , "; this is a simple ini-formatted file"
             , "; lines that start with semi-colons are comments, and removed."
             , "; read `fhc help config` for help on the various options"
             , ";;;;"
             , ""
             , data
             ].concat( [ ";;;;"
                       , "; all options with default values"
                       , ";;;;"
                       ]
                     )
              .concat(Object.keys(dc).map(function (k) {
                return "; " + k + " = " + ini.unParseField(dc[k],k);
              }))
              .concat([""])
              .join("\n");
      fs.writeFile
        ( f
        , data
        , "utf8"
        , function (er) {
            if (er) return cb(er);
            exec("sh", ["-c", e + " "+f], function (er) {
              if (er) return cb(er);
              ini.resolveConfigs(function (er) {
                if (er) return cb(er);
                ini.save(cb);
              });
            });
          }
        );
    });
  });
};

function del (key, cb) {
  if (!key) return cb(new Error("no key provided"));
  ini.del(key);
  ini.save(cb);
};

function set (key, val, cb) {
  if (val === undefined) {
    if (key.indexOf("=") !== -1) {
      var k = key.split("=");
      key = k.shift();
      val = k.join("=");
    } else {
      val = "";
    }
  }
  key = key.trim();
  val = val.trim();
  log("set "+key+" "+val, "config");
  var where = ini.get("global") ? "global" : "user";
  ini.set(key, val, where);
  ini.save(cb);
}

function get (key, cb) {
  if (!key) return list(cb);
  if (key.charAt(0) === "_") {
    return cb(new Error("---sekretz---"));
  }
  var val = fhc.config.get(key);
  output.write(val, function (err) {cb(err,val)});
}

function sort (a, b) {
  return a > b ? 1 : -1;
}

function reverse (a, b) {
  return a > b ? -1 : 1;
}

function list (cb) {
  var msg = ""
    , long = fhc.config.get("long");

  // cli configs.
  // show any that aren't secret
  var cli = ini.configList.list[ini.TRANS.cli]
    , cliKeys = Object.keys(cli).filter(function (k) {
        return !(k.charAt(0) === "_" || types[k] !== types[k]);
      }).sort(function (a, b) {
        return a > b ? 1 : -1;
      });
  if (cliKeys.length) {
    msg += "; cli configs\n";
    cliKeys.forEach(function (k) {
      msg += k + " = " + JSON.stringify(cli[k]) + "\n";
    });
    msg += "\n";
  }

  // env configs
  var env = ini.configList.list[ini.TRANS.env]
    , envKeys = Object.keys(env).filter(function (k) {
        return !(k.charAt(0) === "_" || types[k] !== types[k]);
      }).sort(function (a, b) {
        return a > b ? 1 : -1;
      });
  if (envKeys.length) {
    msg += "; environment configs\n";
    envKeys.forEach(function (k) {
      if (env[k] !== ini.get(k)) {
        if (!long) return
        msg += "; " + k + " = " + JSON.stringify(env[k])
            + " (overridden)\n";
      } else msg += k + " = " + JSON.stringify(env[k]) + "\n";
    });
    msg += "\n";
  }

  // user config file
  var uconf = ini.configList.list[ini.TRANS.user]
    , uconfKeys = Object.keys(uconf).filter(function (k) {
        return types[k] === types[k];
      }).sort(function (a, b) {
        return a > b ? 1 : -1;
      });
  if (uconfKeys.length) {
    msg += "; userconfig " + ini.get("userconfig") + "\n";
    uconfKeys.forEach(function (k) {
      var val = (k.charAt(0) === "_")
              ? "---sekretz---"
              : JSON.stringify(uconf[k]);
      if (uconf[k] !== ini.get(k)) {
        if (!long) return
        msg += "; " + k + " = " + val
            + " (overridden)\n";
      } else msg += k + " = " + val + "\n";
    });
    msg += "\n";
  }

  // global config file
  var gconf = ini.configList.list[ini.TRANS.global]
    , gconfKeys = Object.keys(gconf).filter(function (k) {
        return types[k] === types[k];
      }).sort(function (a, b) {
        return a > b ? 1 : -1;
      });
  if (gconfKeys.length) {
    msg += "; globalconfig " + ini.get("globalconfig") + "\n";
    gconfKeys.forEach(function (k) {
      var val = (k.charAt(0) === "_")
              ? "---sekretz---"
              : JSON.stringify(gconf[k]);
      if (gconf[k] !== ini.get(k)) {
        if (!long) return
        msg += "; " + k + " = " + val
            + " (overridden)\n";
      } else msg += k + " = " + val + "\n";
    });
    msg += "\n";
  }

  // only show defaults if --long
  if (!long) {
    msg += "; node install prefix = " + process.installPrefix + "\n"
         + "; node bin location = " + process.execPath + "\n"
         + "; cwd = " + process.cwd() + "\n"
         + "; HOME = " + process.env.HOME + "\n"
         + "; 'fhc config ls -l' to show all defaults.\n";

    return output.write(msg, cb);
  }

  var defaults = ini.defaultConfig
    , defKeys = Object.keys(defaults);
  msg += "; default values\n";
  defKeys.forEach(function (k) {
    var val = JSON.stringify(defaults[k]);
    if (defaults[k] !== ini.get(k)) {
      if (!long) return;
      msg += "; " + k + " = " + val
          + " (overridden)\n";
    } else msg += k + " = " + val + "\n";
  });
  msg += "\n";

  return output.write(msg, cb);
}

function unknown (action, cb) {
  cb("Usage:\n" + config.usage);
}

config.completion = function (opts, cb) {
  var argv = opts.conf.argv.remain;
  if (argv[1] !== "fhcconfig") argv.unshift("fhcconfig");
  if (argv.length === 2) {
    var cmds = ["get", "set", "delete", "ls", "rm", "edit"];
    if (opts.partialWord !== "l") cmds.push("list");
    return cb(null, cmds);
  }

  var action = argv[2];
  switch (action) {
    case "set":
      // todo: complete with valid values, if possible.
      if (argv.length > 3) return cb(null, []);
      // fallthrough
    case "get":
    case "delete":
    case "rm":
      return cb(null, Object.keys(types));
    case "edit":
    case "list": case "ls":
      return cb(null, []);
    default: return cb(null, []);
  }
};
